### PR TITLE
Features Features Features

### LINKED ISSUE #15: Inconsistent highlight while tapping a tab item
When tapping and holding a tab, the tab uses the `tabBarItem.selectedImage` but the colours don't update:

bildschirmfoto 2015-09-25 um 16 22 55

`JLNRBarController` should do the same as `UITabBar`, namely change the current view controller when TAPPING, not releasing.

### PR DESCRIPTION
**Pull request by sebastianludwig**
Finishes up on #4, implements #1, confirmed #13 and fixed #15.

### CODE REVIEW

**Comment by jlnr** [comment]
Ui, da muss ich gleich doch ein bisschen Zeit blocken, um das zu reviewen :) Top!!

**Comment by jlnr** [comment]
A bit of feedback after playing around with the simulator:
- On/off toggle for animations would be great, just to demonstrate that JLNRBarController looks like UITabController by default
- Dismissing the modal demo VC only makes it show up again. wat? (Probably an old bug I never noticed)
- When double-tapping a tab bar item, no item appears highlighted when the animation finishes

File: Classes/JLNRBackdropSlideAnimator.m [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+@implementation JLNRBackdropSlideAnimator
+
+- (NSTimeInterval)transitionDuration:(id <UIViewControllerContextTransitioning>)transitionContext
+{
+    return 0.8;
+}
+
+- (void)animateTransition:(id<UIViewControllerContextTransitioning>)transitionContext
+{
+    UIViewController* toViewController = [transitionContext viewControllerForKey:UITransitionContextToViewControllerKey];
```
Comment (jlnr) [inline Classes/JLNRBackdropSlideAnimator.m]: Aahhh inconsistent `*` placement aaah!!!

File: Classes/JLNRBackdropSlideAnimator.m [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    } completion:^(BOOL finished) {       // with slow animations turned on in the Simulator, it takes ages before this is called
+        fromView.transform = CGAffineTransformIdentity;
+        toView.transform = CGAffineTransformIdentity;
+        toView.frame = [transitionContext finalFrameForViewController:toViewController];
+        [transitionContext completeTransition:![transitionContext transitionWasCancelled]];
+    }];
+    
+    
+    
+//    // When sliding the views horizontally, in and out, figure out whether we are going left or right.
```
Comment (jlnr) [inline Classes/JLNRBackdropSlideAnimator.m]: Is this block of code still relevant?

File: Classes/JLNRBadgeView.m [hunk 3]
```diff
@@ -55,8 +58,8 @@ - (void)drawRect:(CGRect)rect
 - (CGSize)intrinsicContentSize
 {
     CGSize labelSize = _label.intrinsicContentSize;
-    CGFloat height = labelSize.height + 2 * kVerticalPadding;
-    return CGSizeMake(MAX(labelSize.width + 3 * kHorizontalPadding, height), height);
+    CGFloat height = labelSize.height + 2 * self.verticalPadding;
+    return CGSizeMake(MAX(labelSize.width + 3 * self.horizontalPadding, height), height);
 }
 
 - (NSString *)bagdeText
```
Comment (jlnr) [inline Classes/JLNRBadgeView.m]: BTW you are using Swift too much recently, in `setBadgeText:` you check for `nil` :P

File: Classes/JLNRBadgeView.h [hunk 4]
```diff
@@ -15,5 +15,10 @@ IB_DESIGNABLE
 @property (nonatomic) IBInspectable UIColor *textColor;
 @property (nonatomic) IBInspectable UIColor *badgeColor;
 
+@property (nonatomic) IBInspectable UIFont *font;
```
Comment (jlnr) [inline Classes/JLNRBadgeView.h]: `IB_DESIGNABLE` & `IBInspectable` - do you think it makes sense to make this a publicly reusable class?

File: Classes/JLNRBarController.h [hunk 5]
```diff
@@ -11,11 +11,16 @@
 // Let this header function as an umbrella header.
 #import "JLNRBarCell.h"
 #import "JLNRBarView.h"
+#import "JLNRBackdropSlideAnimator.h"
```
Comment (jlnr) [inline Classes/JLNRBarController.h]: Same question here, just to confirm - this will be a publicly reusable class?

File: Classes/JLNRBarCell.m [hunk 6]
```diff
@@ -59,6 +59,9 @@ - (void)createMenuCellSubviews
     JLNRBadgeView *badgeView = [[JLNRBadgeView alloc] initWithFrame:CGRectMake(0, 0, 30, 20)];
     badgeView.badgeColor = [UIColor redColor];
     badgeView.textColor = [UIColor whiteColor];
+    badgeView.horizontalPadding = 2;
+    badgeView.verticalPadding = 1;
+    badgeView.font = [UIFont systemFontOfSize:13];
```
Comment (jlnr) [inline Classes/JLNRBarCell.m]: JLNRBadgeView sets `.font` to `sytemFontOfSize:14` in its constructor, then lets `JLNRBarCell` override it - which is the right size?

File: Classes/JLNRBadgeView.h (hunk 4 above)
Comment (jlnr) [inline Classes/JLNRBadgeView.h]: Also, I kind of wish this worked with UIAppearance. But I have to admit that debugging UIAppearance is a little confusing...

File: Classes/JLNRBarController.m [hunk 7]
```diff
@@ -15,6 +16,9 @@ @interface JLNRBarController () <JLNRBarViewDelegate>
 
 
 @implementation JLNRBarController
+{
+    BOOL _transitionInProgress;
```
Comment (jlnr) [inline Classes/JLNRBarController.m]: If you already have the anonymous category right above the `@implementation`, then I think should should be a `@property`

**Comment by jlnr** [comment]
While I am leaving nit-picky comments, two higher-level thoughts.

First, what about allowing `selectedIndex == -1` or `== NSNotFound` as a "nothing is selected" state. Maybe that can clear up the `tabHasEverBeenSelected` magic? I am not *really* sure if this is a good idea, but I know that the App Store app unselects all tabs if there is no internet connection, and displays an error message in the content area.

Second, the transition delegate is pretty sweet! It still breaks in landscape mode (iPhone 6s Plus simulator). Do you think this animation is good enough to guide users through tab changes in $OurAwesomeApp? I think it'd be nice if the selected tab's background (`selectionIndicatorColor`) could move from one tab to another, but I don't see how we could possibly write an API that allows for generic tab bar animations. Should I open a ticket to add this as a special-case feature? (I guess I'll only use it for $ClientApp, anyway.)

And yet another nitpick: I think that setting `selectedIndex` in code should not trigger the delegate method, only user interaction should. I can open another ticket for that, anyway.

**Comment by jlnr** [comment]
BTW rotating the simulator while the animation is playing also breaks things :) I hate GUI code.

File: Classes/JLNRBadgeView.h (hunk 4 above)
Comment (sebastianludwig) [inline Classes/JLNRBadgeView.h]: You know where these attributes came from ;-) I'm not completely sure if it makes sense to extract this class into yet another badge pod..?

File: Classes/JLNRBarController.h (hunk 5 above)
Comment (sebastianludwig) [inline Classes/JLNRBarController.h]: Yes. Like it's used in the Example at the moment. So one is able to use a meaningful jump animation right out of the box. (BTW, I'm not very happy with the way it looks at the moment.)

File: Classes/JLNRBarCell.m (hunk 6 above)
Comment (sebastianludwig) [inline Classes/JLNRBarCell.m]: In this project the `JLNRBarCell` sets the correct size for tab bar like badges. In the project where I ripped `JLNRBadgeView` out, the `systemFontOfSize:14` is correct (aka default).

**Comment by sebastianludwig** [comment]
And it's not getting better with what I'm working on right now :-/

**Comment by sebastianludwig** [comment]
Could you open three tickets for the higher level issues? Thread mixing is too exhausting ;-)

**Comment by sebastianludwig** [comment]
> Dismissing the modal demo VC only makes it show up again. wat? (Probably an old bug I never noticed)

That's a temporary feature (see 53cc256). I'm also not sure if presenting the bar controller modally should be the default example. It's probably a rather odd use case.

File: Classes/JLNRBadgeView.h (hunk 4 above)
Comment (jlnr) [inline Classes/JLNRBadgeView.h]: No, not a pod, but a documented part of LDOBarController :)

Comment (sebastianludwig) [inline Classes/JLNRBadgeView.h]: A pod would be cleaner though. Objections?

Comment (jlnr) [inline Classes/JLNRBadgeView.h]: I'd say we should wait until we are working on an app that requires `*BadgeView` but does not use `*BarController`.

Comment (jlnr) [inline Classes/JLNRBadgeView.h]: (Your To Do list app, naturally ;) )

**Pull request merged**